### ISSUE TITLE
Rational Quadratic kernel

### ISSUE THREAD

**Issue by refrigerator** [GPflow/GPflow#526]
Hi folks,

Love the library - thanks for all your hard work. Any plans to add a Rational Quadratic kernel? If not, I'd like to add it myself, but just wanted to double check - I can plausibly do it in a couple of hours by following http://gpflow.readthedocs.io/en/latest/notebooks/kernels.html#Making-new-kernels right? Or am I overlooking something that means it'll take considerably longer?

Thanks

Taimur

**Comment by mathDR** [comment]
What about something like:
```
class RationalQuadratic(GPflow.kernels.Stationary):
    def __init__(self):
        GPflow.kernels.Stationary.__init__(self, input_dim=1, active_dims=[0])
        self.alpha = GPflow.param.Param(1.0, transform=GPflow.transforms.positive)

    def K(self, X, X2=None, presliced=False):
        if not presliced:
            X, X2 = self._slice(X, X2)
        return self.variance * tf.pow(1 + -self.square_dist(X, X2) / (2*self.alpha),-self.alpha)
```

**Comment by refrigerator** [comment]
@mathDR that looks good to me! 👍

**Comment by refrigerator** [comment]
Is adding this really as simple as adding that to the `kernels.py` file? Are there any unit tests or anything like that that we need to write?

**Comment by refrigerator** [comment]
I tried running the code above but getting this:

`InvalidArgumentError (see above for traceback): Cholesky decomposition was not successful. The input might not be valid.
	 [[Node: model.build_likelihood/conditional/Cholesky = Cholesky[T=DT_DOUBLE, _device="/job:localhost/replica:0/task:0/device:CPU:0"](model.build_likelihood/conditional/add)]]`

**Comment by mathDR** [comment]
What's your input data?

**Comment by mathDR** [comment]
@refrigerator Did you get this to work?

**Comment by shz9** [comment]
I got the same error as @refrigerator, and what fixed it for me was removing the minus sign before the square distance. My implementation:

```
def K(self, X, X2=None, presliced=False):
        if not presliced:
            X, X2 = self._slice(X, X2)
        return self.variance * tf.pow(1 + self.square_dist(X, X2) / (2*self.alpha*(self.lengthscales ** 2)),
                                      -1 * self.alpha)
```

I'm experimenting with the [Mauna Loa CO2](http://mldata.org/repository/data/viewslug/mauna-loa-atmospheric-co2/) dataset.

**Comment by Joshuaalbert** [comment]
@shz9 any followup?

**Comment by shz9** [comment]
My implementation seems to work. Here's the full class definition:

```
class RationalQuadratic(gpflow.kernels.Stationary):
    """
    This class extends the Stationary kernel in GPflow and
    implements the Rational Quadratic kernel.
    Based on:
    https://github.com/GPflow/GPflow/issues/526#issuecomment-338256036
    """
    def __init__(self, input_dim, active_dims=None, alpha=1.0, lengthscales=1.0):
        gpflow.kernels.Stationary.__init__(self, input_dim=input_dim, active_dims=active_dims)
        self.alpha = gpflow.param.Param(alpha, transform=gpflow.transforms.positive)
        self.lengthscales = lengthscales

    def K(self, X, X2=None, presliced=False):
        if not presliced:
            X, X2 = self._slice(X, X2)
        return self.variance * tf.pow(1 + (self.square_dist(X, X2) / (2*self.alpha*(self.lengthscales ** 2))),
                                      -1 * self.alpha)
```

The code [here](https://github.com/shz9/csc2541-ml-project/blob/master/GP_experiments/kernels/CustomKernels.py) shows how I used the implementation above to replicate the Mauna Loa CO2 example.

**Comment by refrigerator** [comment]
nice one @shz9 !

**Comment by st--** [comment]
@shz9 I think your implementation might actually scale the distances by the lengthscales twice (once inside self.square_dist; this was confusing, it's now renamed to self.scaled_square_dist...).

@refrigerator I hadn't noticed this PR before but also needed a RationalQuadratic kernel a while ago, and this is now merged in master (#666).
